### PR TITLE
Add Redis Ring.

### PR DESCRIPTION
**Pull request by vmihailenco**
I called it Ring, because client uses consistent hash ring to distribute keys.

@dim would be nice to hear you opinion.

### CODE REVIEW

**Comment by dim** [comment]
@vmihailenco will try to look through today/tomorrow. I get the idea, but wouldn't it make more sense to wrap this into a separate package/repo? It's not a core redis concept, especially not since cluster support was added. It also introduces a new dependency, which is not the end of the world, but a new package seems just cleaner overall.

**Comment by vmihailenco** [comment]
I am fine with separate package, but that requires exporting `process` method and doing something about `cl.baseClient.process` ambiguity.

**Comment by vmihailenco** [comment]
@dim ping.

Even though `process` is already exported I would like to keep Ring in this repo. Probably let's move it to subpackage?

**Comment by vmihailenco** [comment]
Moving to subpackage requires exporting `Commandable` struct, which I don't want to do (we did not have that struct in v2 and there are chances we will not have it in v4). Going to merge this as is tomorrow if I don't hear anything from you :/

File: ring.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+	if err != nil {
+		cmd.setErr(err)
+		return
+	}
+	cl.baseClient.process(cmd)
+}
+
+// rebalance removes dead shards from the ring.
+func (ring *Ring) rebalance() {
+	defer ring.mx.Unlock()
```
Comment (dim) [inline ring.go]: Again :)

File: ring.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+		PoolSize:    opt.PoolSize,
+		PoolTimeout: opt.PoolTimeout,
+		IdleTimeout: opt.IdleTimeout,
+	}
+}
+
+type ringShard struct {
+	Client *Client
+	down   int
```
Comment (dim) [inline ring.go]: Checked briefly, looks OK. ringShards don't need to be thread-safe since they are guarded by `mx` in Ring

File: ring.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
+package redis
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/golang/groupcache/consistenthash"
```
Comment (dim) [inline ring.go]: Any change you could copy and paste this across? Looks like it's short enough to be embedded. I prefer not to have external dependencies, especially when it comes to hashing.

**Comment by dim** [comment]
Really sorry about the delay, feel free to merge. The only (minor) concerns I have are:

1. importing hashing algorithms seems risky, I would prefer to embed it to be *really* consistent
2. how relevant are ring implementations, given that redis cluster is now available?

File: ring.go (hunk 3 above)
Comment (vmihailenco) [inline ring.go]: Will do.

**Comment by vmihailenco** [comment]
>how relevant are ring implementations, given that redis cluster is now available?

Redis Cluster requires using replication and slaves so it can do failover when master dies. That is a bit expensive for us atm. So it is relevant at least for me :)

**Comment by dim** [comment]
Fair point

**Comment by vmihailenco** [comment]
Done. Package `consistenthash` was copied to internal folder.

**Pull request merged**